File: haskell/src/Solutions/Euler033.hs
module Solutions.Euler033 where

prod' (a, b) (c, d) = (a * c, b * d)
ans (num, den) = den `div` gcd den num

compute = ans $ foldr1 prod' [(a, b) |
    a <- [11..99],
    a `mod` 10 /= 0,
    b <- [a + 1..99],
    b `mod` 10 /= 0,
    a `mod` 10 == b `div` 10,
    a * (b `mod` 10) == b * (a `div` 10)]

euler033 :: IO String
euler033 = return $ show compute

File: rust/src/solutions/euler033.rs
pub fn euler033() -> String {
    let mut num = 1;
    let mut den = 1;
    for a in 11..100 {
        for b in a + 1..100 {
            if b % 10 != 0 && a % 10 == b / 10 && a * (b % 10) == b * (a / 10) {
                num *= a;
                den *= b;
            }
        }
    }
    (den / num::integer::gcd(den, num)).to_string()
}
